General frontend refactor

sbracellari opened the pull request:
#### Overview

Since courses backend was rewritten, we figured it was time to pull in the most recent version of courses frontend. However, since the backend changed quite a bit in the portlet to soffit conversion, some reworking of the frontend code was necessary to get everything working properly.

This is a really large pull request (my bad), and I've been looking at this code for a long time, so I'm sure that I've missed stuff. I'd appreciate if you guys could really dig into this PR, if you have the time.

You can either run the project locally by changing the `is_demo` variable in the `index.html` to `true` or by running courses backend and frontend simultaneously. I would suggest the latter, as it would ensure all functionality (printing especially) works as expected. I've tested on chrome and firefox. If anyone can test on edge or safari, it would be greatly appreciated.

A general list of what's changed: 

- Anything advising related has been ripped out
- Reducers/actions have been reworked to fit new endpoint structure
- The react calendar has been replaced with a third party option, Full Calendar (sorry Aaron!)
- The local backend has been removed and replaced with demo JSON data located in `src/api`
- Translations was removed as it was outdated and we are planning to look into a new way of handling that
- `prettier` was added to the project and run
- Unused/unnecessary code was removed and the project was cleaned up generally

#### Merger
- [x] @amgoodfellow 

#### Reviewers
- [x] @richast
- [x] @sterlingwalker
- [x] @nickwoodle
- [x] @kdjohnson 

**NOTE:** The amount of additions/deletions looks HUGE, I mainly blame this on `prettier`. Also, you'll want to run this branch with courses backend branch CS-9 (until that is merged in). Those two work together

For running w demo data, change `is_demo` in the `index.html` to true, and pass `true` instead of `false` to the `generate_pdf` function called in `src/components/PrintCourses.js` (forgot to pull `is_demo` in for that one. will be fixed in the next PR)

sbracellari committed:
removes anything advising related #209

the courses soffit doesn't have advising baked in anymore, so anything
advising related in the frontend is no longer needed

sbracellari committed:
removes excess reducers/actions #209

the soffit pulls data in differently, so some actions/reducers
ended up not being needed in the end

sbracellari committed:
removes any use of useTranslation #209

translations breaks in portlet to soffit conversions, so its getting
ripped out for now, until we can find a better way to do it. other
commits might also include translations stuff

sbracellari committed:
changes buildingRoom to location #209

sbracellari committed:
updates reducers, actions, and api.js #209

sbracellari committed:
adds pdf service #209

the pdf service was ripped out of the backend. this is its replacement

sbracellari committed:
general refactor #209

commits the rest of the general refactor, where most of the state
changes are happening

sbracellari committed:
updates deps #209

sbracellari committed:
modifications to print element #209

sbracellari committed:
improves styling and flow of state #209

sbracellari committed:
updates error card & adds it to grade section #209

sbracellari committed:
fix styling bugs/remove unused props #209

sbracellari committed:
replaces local backend with demo data #209

sbracellari committed:
changes to 3rd party calendar, modify actions #209

sbracellari committed:
misc styling/cleaning up #209

sbracellari committed:
updates pdf service #209

pdf generation has been moved back to the backend

sbracellari committed:
fix anchorEl bug in popover #209

sbracellari committed:
add/run prettier, remove unused code #209

sbracellari committed:
updates readme, turns demo data off #209

sterlingwalker commented on src/components/Courses.js:
```diff
@@ -1,149 +1,128 @@
+        <CircularProgress color='secondary' size={50} />
+      </div>
+    )
+  } else if (courses_fetched === true && courses_error === true) {
```
Since these are boolean values you won't need to write out the full `=== true` each time. There are a few other places where these expressions could be shortened.

sbracellari replied:
True. Some of this code could have been copy/pasted from other areas or I was just writing it to match what was already there. But I can go through and shorten things where possible. Thanks!

sbracellari replied:
Going to address this in #212

sterlingwalker reviewed (commented):
Great Sam job taking on such a large conversion project! The only other thing I would have recommended was simplifying some of the longer components such as `Instructors.js` but its not that important right now.

sbracellari commented:
@sterlingwalker you have a lot of good suggestions, and they've got me looking at other areas in the code that I can update as well.

I think that I'm going to opt to only add in bug fixes in this merge request since its already huge. But I'm going to branch off of this branch and continue working on code enhancements. How's that sound?

sterlingwalker commented:
> @sterlingwalker you have a lot of good suggestions, and they've got me looking at other areas in the code that I can update as well.
> 
> I think that I'm going to opt to only add in bug fixes in this merge request since its already huge. But I'm going to branch off of this branch and continue working on code enhancements. How's that sound?

That works for me! I'll check my box.

sbracellari commented on src/components/WaitlistCourse.js:
```diff
@@ -1,90 +1,57 @@
+  },
 }))
 
 export default function WaitlistCourse(props) {
```
Going to address this in #212

sbracellari commented on src/components/TermSelect.js:
```diff
@@ -1,150 +1,108 @@
+    terms.forEach((term) => {
+      if (term.code === event.target.value) {
+        dispatch(update_term(term))
+        setSelectedTerm(term)
       }
-
-      dispatch(fetch_courses(new_term))
-      dispatch(set_current_term(new_term))
-    }
+    })
   }
```
Going to address this in #212

sbracellari committed:
updates demo data per @nickwoodle suggestion #209

also patches a bug found after changing a class's demo data to have
no meetings

nickwoodle reviewed (commented):
Sorry to be a pain. I might drop the `@oakland.edu` and just put something like `@email.test` or something that we know is not possible to be someone's email.

sbracellari committed:
replaces oakland domain with test domain #209

sbracellari commented on src/api/courses.json:
```diff
@@ -0,0 +1,420 @@
+                    "firstName": "Hany",
+                    "lastName": "Othman",
+                    "office": "N/A",
+                    "email": "test.othman@oakland.edu"
```
This is all up to date with fake names and test domains!

sbracellari committed:
change pdf endpoint wording #209

the endpoint changed in the backend so i have to change it here too

Richast commented:
LGTM :100: 

I really like these changes, everything is much `prettier` now. Worked with the backend find and I didn't find any bugs.

nickwoodle commented:
LGTM! I am good with it!

amgoodfellow commented on package.json:
```diff
@@ -2,9 +2,15 @@
+    "@fullcalendar/timegrid": "^5.1.0",
     "@material-ui/core": "^4.0.0",
     "@material-ui/icons": "^4.0.0",
+    "@react-pdf/renderer": "^2.0.0-beta.14",
```
Are we using the react-pdf renderer?

sbracellari replied:
nope. will remove that

sbracellari replied:
this was removed along with i18 deps, since those aren't being used anymore either

amgoodfellow commented on package.json:
```diff
@@ -25,14 +31,20 @@
+    "eject": "react-scripts eject",
+    "format": "npx prettier --write \"src/**/*.js\""
+  },
+  "prettier": {
```
This is a nice addition to the project. I know that our most recent change with the `"printWidth": 100` came after this PR, but maybe it could be added in now as well? Since there's already a lot changed because of prettier formatting anyway...

sbracellari replied:
fair point. i added it in #212 but i can add it here

amgoodfellow commented on README.md:
```diff
@@ -1,46 +1,9 @@
-
-## Modifying data
-Note if you ever want to change data in you table. Do the following `update TABLE set COLUMN = VALUE where id = ID_NUMBER_OF_ROW;`
+To run this project locally (with no backend), navigate to `public/index.html` and change `is_demo` to `true`. This will provide the frontend with realistic course and calendar data.
```
Thank you for updating the README! That's so easy to skip over. And it'd be confusing for future people to see all this stuff about Go when there's not `.go` files in the project anymore lol

amgoodfellow commented on src/Calendar.css:
```diff
@@ -0,0 +1,55 @@
+}
+
+.fc .fc-button-primary:not(:disabled):active, .fc .fc-button-primary:not(:disabled).fc-button-active {
+  background-color: #dbeeff;
```
I love this hex value -- where's dbeeff?

amgoodfellow commented on src/actions/coursesActions.js:
```diff
@@ -1,22 +1,18 @@
-    get_courses(new_term, courses_url)
-      .then(courses => {
-        dispatch({ type: 'RECEIVE_COURSES', payload: courses })
+    get_courses(is_demo, term, courses_url)
```
I actually don't know how other people feel about this. But. 

In some other projects, `IS_DEMO` is defined as a constant in the `public/index.html` file and then in the soffit's `view.jsp`. Then in the `app.js` file it's just a global variable. The function arguments don't include `is_demo`, but the check still happens based on the global variable being pulled in. I don't know if people have a preference either way, but it might be nice to get some feedback so all our repos can do it the same way

sbracellari replied:
maybe we can bring this up in our standards meeting on the 31st?

amgoodfellow replied:
That sounds good to me

amgoodfellow commented on src/actions/eventsActions.js:
```diff
@@ -0,0 +1,18 @@
+
+export function fetch_events(term) {
+  return function (dispatch) {
+    dispatch({ type: 'FETCH_EVENTS_START', payload: {} })
```
I like the flow for the fetching -> receive/error

amgoodfellow commented on src/api/courses.json:
```diff
@@ -0,0 +1,445 @@
+            "instructors": [
+                {
+                    "crn": "43069",
+                    "firstName": "Iraida",
```
Maybe we should replace these guys with demo data too

sbracellari replied:
the names are demo data. i used the name generator website you suggested lol

amgoodfellow replied:
Ha! Nice - they seemed so realistic ;)

amgoodfellow commented on src/components/CourseDetails.js:
```diff
@@ -60,41 +49,41 @@ export default function CourseDetails(props) {
+          aria-label='course description'
         >
-          {t('description')}
+          Description
```
This is a much bigger scope than this PR, but a fun upcoming task for anyone who wants it is going to be getting translations back into the portal in a more reliable way

amgoodfellow commented on src/components/Courses.js:
```diff
@@ -1,149 +1,128 @@
-                className={classes.content}
+  } else {
+    let elements = []
+    for (let [i, course] of courses.entries()) {
```
Nice! I really like the `.entries()` - I actually wasn't aware of that method's existence 

Maybe eventually the different methods of iteration across the components can be made a bit more similar? This is a `for` loop on entries, other places do some explicit indexing based on sizes of arrays, other places do `foreach`

sbracellari replied:
i don't wanna burst your bubble, but i've gotten rid of pretty much all of the for loops in #212... lol. i switched to maps for everything

amgoodfellow replied:
Oh no that's fine. @kdjohnson can tell you - I was very against the use of `.map()` when I started at UTS, but I think they are my favorite sort of iteration now

amgoodfellow commented on src/components/Courses.js:
```diff
@@ -1,149 +1,128 @@
+        )
+      } else if (!Object.is(course.waitlist, '0')) {
+        elements.push(
+          <WaitlistCourse course={course} key={'waitlist' + Math.random()} />
```
Also, these `Math.random()` boyos. Maybe we tried this before and it wasn't unique enough, but could we do something like ```key={`waitlist-${course.crn}`}```?  It does work as-is, but it feels a bit hacky. And I know it was like that before, so I don't think it'll hold up this PR. But maybe as a follow-up issue

sbracellari replied:
i think i've also gotten rid of this in #212, a lot of components were combined so there wasn't the need for a ton of super unique keys. i'm using maps for everything so i'm just using the index as the key, and there hasn't seemed to be any console errors regarding duplicate/missing keys

amgoodfellow replied:
Cool, that sounds like a good solution to me :+1:

amgoodfellow commented on src/components/ExpandableMeetings.js:
```diff
@@ -74,37 +75,41 @@ export default function ExpandableMeetings(props) {
-            rel="noopener noreferrer"
+            tabIndex='0'
+            target='_blank'
+            href={getMapUrl(meeting.location, false)}
```
Ooh. Also something related to this repo but not this exact PR. It might be a good task for someone to go through and update the mapUrls - especially if the new dorm is on there now

sbracellari replied:
True that

amgoodfellow commented on src/reducers/index.js:
```diff
@@ -1,12 +1,103 @@
+    case 'FETCH_EVENTS_START':
+    case 'FETCH_COURSES_START':
+    case 'FETCH_SELECTED_COURSES_START': {
+      return { ...state, fetching: true, fetched: false, error: false }
```
Interesting - I can't remember, were all of the fetching_start actions always grouped into the same case statement?

It'd seem to me like if a particular fetch failed or we wanted to display a loading icon for just the course content, we wouldn't know which particular fetch we were talking about. I guess if it was always like this then it's something that we haven't needed yet

sbracellari replied:
i don't think it's all like this, i just did it this way. i can separate them if you would like, tho

sbracellari replied:
i was just trying to avoid adding a bunch of different state variables

sbracellari replied:
@amgoodfellow, would you prefer I fix this here or in #212?

amgoodfellow replied:
I think it's fine to go in #212

amgoodfellow reviewed (approved):
This looks great @sbracellari! Lots of work was obviously put in. A lot of stuff was simplified, particularly with the removal of advisor views. 

I don't think any of my comments were really show-stoppers for this merge - lots could be dismissed or addressed in a follow-up issue. Or maybe just discussed. 

Excited to merge this into OU's fork :+1:

kdjohnson commented on src/components/CoursesTabs.js:
```diff
@@ -1,114 +1,127 @@
+  const selected_term = useSelector((state) => state.selected_term)
   const classes = useStyles()
+  const theme = useTheme()
+  const mobile = useMediaQuery(theme.breakpoints.down('xs'))
```
This is such a better way of determining this.

sbracellari replied:
You can thank @amgoodfellow for that one

kdjohnson commented on src/components/Grades.js:
```diff
@@ -88,75 +78,78 @@ const getOverallCredits = credits => {
-              </TableHead>
-              <TableBody>{getOverallCredits(credits)}</TableBody>
-            </Table>
+  if (credits !== []) {
```
Is this if check for type? If not is this needed even though you check the length of credits on [line 89](f563e617b0ac9a67390d725a3d3bb6c2R89)?

sbracellari replied:
This check isn't needed. Things are being done a bit different in #212 where I'm doing some more refactoring, so I think I'll just leave this as is for this PR if that's okay with you

amgoodfellow replied:
I think that'd be fine with KaJuan lol. I believe he messaged me after his reviewing saying things looked good to him. Addressing this in #212 seems reasonable

sbracellari committed:
removes unneeded deps #209
